src/Microsoft.ApplicationInsights.AspNetCore/TelemetryInitializers/AuthenticatedUserIdTelemetryInitializer.cs

```diff
@@ -0,0 +1,45 @@
+﻿namespace Microsoft.ApplicationInsights.AspNetCore.TelemetryInitializers
+{
+    using Microsoft.ApplicationInsights.AspNetCore.Extensions;
+    using Microsoft.ApplicationInsights.Channel;
+    using Microsoft.ApplicationInsights.DataContracts;
+    using Microsoft.AspNetCore.Http;
+    using Microsoft.Extensions.Options;
+
+    /// <summary>
+    /// This initializer sets the Authenticated User Id if enabled via <see cref="ApplicationInsightsServiceOptions.EnableAuthenticationTracking"/>.
+    /// </summary>
+    public class AuthenticatedUserIdTelemetryInitializer : TelemetryInitializerBase
+    {
+        private readonly bool enabled;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AuthenticatedUserIdTelemetryInitializer" /> class.
+        /// </summary>
+        /// <param name="options">Provides the option <see cref="ApplicationInsightsServiceOptions.EnableAuthenticationTracking"/> indicating whether the currently authenticated user is added to telemetries or not.</param>
+        /// <param name="httpContextAccessor">Accessor to provide HttpContext corresponding to telemetry items.</param>
+        public AuthenticatedUserIdTelemetryInitializer(IOptions<ApplicationInsightsServiceOptions> options, IHttpContextAccessor httpContextAccessor)
+             : base(httpContextAccessor)
+        {
+            this.enabled = options.Value.EnableAuthenticationTracking;
+        }
+
+        /// <inheritdoc />
+        protected override void OnInitializeTelemetry(HttpContext platformContext, RequestTelemetry requestTelemetry, ITelemetry telemetry)
+        {
+            if (this.enabled)
```

Comment:
I'm not fully satisfied with this solution of always registering the `AuthenticatedUserIdTelemetryInitializer` but only applying its logic if `EnableAuthenticationTracking`/`enabled` is set.

I would've preferred not to register the telemetry initializer at all if `EnableAuthenticationTracking` is not enabled.

However, unfortunately, `ApplicationInsightsServiceOptions` are not available / hard to access in `ApplicationInsightsExtensions.AddApplicationInsightsTelemetry()`. That's why I went for the approach using the `enabled` flag.

Reply:
For now, i think this is the best you can do.